<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Подключение-библиотек" data-toc-modified-id="Подключение-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Подключение библиотек</a></span></li><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Функция-первичного-анализа-данных" data-toc-modified-id="Функция-первичного-анализа-данных-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Функция первичного анализа данных</a></span></li><li><span><a href="#Первичный-анализ-данных" data-toc-modified-id="Первичный-анализ-данных-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Первичный анализ данных</a></span></li><li><span><a href="#Проверка-на-дубликаты" data-toc-modified-id="Проверка-на-дубликаты-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Проверка на дубликаты</a></span></li><li><span><a href="#Проверка-пропущенных-значений" data-toc-modified-id="Проверка-пропущенных-значений-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Проверка пропущенных значений</a></span></li><li><span><a href="#Проверка-значений-столбца-Geography" data-toc-modified-id="Проверка-значений-столбца-Geography-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>Проверка значений столбца <code>Geography</code></a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.8"><span class="toc-item-num">1.8&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Изучение-баланса-классов-целевого-признака" data-toc-modified-id="Изучение-баланса-классов-целевого-признака-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Изучение баланса классов целевого признака</a></span></li><li><span><a href="#Формирование-признаков-и-целевого-признака" data-toc-modified-id="Формирование-признаков-и-целевого-признака-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Формирование признаков и целевого признака</a></span></li><li><span><a href="#Преобразование-категориальных-признаков,-а-так-же-масштабирование-признаков" data-toc-modified-id="Преобразование-категориальных-признаков,-а-так-же-масштабирование-признаков-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Преобразование категориальных признаков, а так же масштабирование признаков</a></span></li><li><span><a href="#Разбивка-данных-на-обучающую,-валидационную-и-тестовую-выборки" data-toc-modified-id="Разбивка-данных-на-обучающую,-валидационную-и-тестовую-выборки-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Разбивка данных на обучающую, валидационную и тестовую выборки</a></span></li><li><span><a href="#Функция-подбора-гиперпараметров-для-модели-дерево-решений" data-toc-modified-id="Функция-подбора-гиперпараметров-для-модели-дерево-решений-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Функция подбора гиперпараметров для модели дерево решений</a></span></li><li><span><a href="#Сохранение-модели-дерево-решений-с-наилучшими-параметрами" data-toc-modified-id="Сохранение-модели-дерево-решений-с-наилучшими-параметрами-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Сохранение модели дерево решений с наилучшими параметрами</a></span></li><li><span><a href="#Функция-подбора-гиперпараметров-для-модели-слуйчайный-лес" data-toc-modified-id="Функция-подбора-гиперпараметров-для-модели-слуйчайный-лес-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Функция подбора гиперпараметров для модели слуйчайный лес</a></span></li><li><span><a href="#Сохранение-модели-случайный-лес-с-наилучшими-гиперпараметрами" data-toc-modified-id="Сохранение-модели-случайный-лес-с-наилучшими-гиперпараметрами-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>Сохранение модели случайный лес с наилучшими гиперпараметрами</a></span></li><li><span><a href="#Функция-для-подбора-гиперпараметров-для-модели-логистическая-регрессия" data-toc-modified-id="Функция-для-подбора-гиперпараметров-для-модели-логистическая-регрессия-2.9"><span class="toc-item-num">2.9&nbsp;&nbsp;</span>Функция для подбора гиперпараметров для модели логистическая регрессия</a></span></li><li><span><a href="#Сохранение-модели-логистическая-регрессия" data-toc-modified-id="Сохранение-модели-логистическая-регрессия-2.10"><span class="toc-item-num">2.10&nbsp;&nbsp;</span>Сохранение модели логистическая регрессия</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.11"><span class="toc-item-num">2.11&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Взвешивание-классов" data-toc-modified-id="Взвешивание-классов-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Взвешивание классов</a></span><ul class="toc-item"><li><span><a href="#Модель-дерево-решений" data-toc-modified-id="Модель-дерево-решений-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Модель дерево решений</a></span></li><li><span><a href="#Модель-случайный-лес" data-toc-modified-id="Модель-случайный-лес-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Модель случайный лес</a></span></li><li><span><a href="#Модель-логистической-регрессии" data-toc-modified-id="Модель-логистической-регрессии-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Модель логистической регрессии</a></span></li></ul></li><li><span><a href="#Увеличение-выборки" data-toc-modified-id="Увеличение-выборки-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Увеличение выборки</a></span><ul class="toc-item"><li><span><a href="#Функция-увеличения-обучающей-выборки" data-toc-modified-id="Функция-увеличения-обучающей-выборки-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Функция увеличения обучающей выборки</a></span></li><li><span><a href="#Сохранение-увеличенных-выборок" data-toc-modified-id="Сохранение-увеличенных-выборок-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Сохранение увеличенных выборок</a></span></li><li><span><a href="#Модель-дерево-решений" data-toc-modified-id="Модель-дерево-решений-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Модель дерево решений</a></span></li><li><span><a href="#Модель-случайный-лес" data-toc-modified-id="Модель-случайный-лес-3.2.4"><span class="toc-item-num">3.2.4&nbsp;&nbsp;</span>Модель случайный лес</a></span></li><li><span><a href="#Модель-логистическая-регрессия" data-toc-modified-id="Модель-логистическая-регрессия-3.2.5"><span class="toc-item-num">3.2.5&nbsp;&nbsp;</span>Модель логистическая регрессия</a></span></li></ul></li><li><span><a href="#Уменьшение-выборки" data-toc-modified-id="Уменьшение-выборки-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Уменьшение выборки</a></span><ul class="toc-item"><li><span><a href="#Функция-уменьшения-обучающей-выборки" data-toc-modified-id="Функция-уменьшения-обучающей-выборки-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Функция уменьшения обучающей выборки</a></span></li><li><span><a href="#Сохранение-уменьшенных-выборок" data-toc-modified-id="Сохранение-уменьшенных-выборок-3.3.2"><span class="toc-item-num">3.3.2&nbsp;&nbsp;</span>Сохранение уменьшенных выборок</a></span></li><li><span><a href="#Модель-дерево-решений" data-toc-modified-id="Модель-дерево-решений-3.3.3"><span class="toc-item-num">3.3.3&nbsp;&nbsp;</span>Модель дерево решений</a></span></li><li><span><a href="#Модель-случайный-лес" data-toc-modified-id="Модель-случайный-лес-3.3.4"><span class="toc-item-num">3.3.4&nbsp;&nbsp;</span>Модель случайный лес</a></span></li><li><span><a href="#Модель-логистическая-регрессия" data-toc-modified-id="Модель-логистическая-регрессия-3.3.5"><span class="toc-item-num">3.3.5&nbsp;&nbsp;</span>Модель логистическая регрессия</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span><ul class="toc-item"><li><span><a href="#Функция-подсчета-площади-под-ROC-кривой" data-toc-modified-id="Функция-подсчета-площади-под-ROC-кривой-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Функция подсчета площади под ROC-кривой</a></span></li><li><span><a href="#Модель-дерево-решений" data-toc-modified-id="Модель-дерево-решений-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Модель дерево решений</a></span></li><li><span><a href="#Модель-случайный-лес" data-toc-modified-id="Модель-случайный-лес-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Модель случайный лес</a></span></li><li><span><a href="#Модель-логистическая-регрессия" data-toc-modified-id="Модель-логистическая-регрессия-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Модель логистическая регрессия</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Необходимо спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Для обучения моделей предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Допустимые значения `F1-меры` для выбранной модели прогнозирования - не ниже 0.59

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

# Описание данных

**Признаки**

* `RowNumber` — индекс строки в данных
* `CustomerId` — уникальный идентификатор клиента
* `Surname` — фамилия
* `CreditScore` — кредитный рейтинг
* `Geography` — страна проживания
* `Gender` — пол
* `Age` — возраст
* `Tenure` — сколько лет человек является клиентом банка
* `Balance` — баланс на счёте
* `NumOfProducts` — количество продуктов банка, используемых клиентом
* `HasCrCard` — наличие кредитной карты
* `IsActiveMember` — активность клиента
* `EstimatedSalary` — предполагаемая зарплата

**Целевой признак**

* `Exited` — факт ухода клиента

## Подготовка данных

### Подключение библиотек

In [1]:
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
pd.options.mode.chained_assignment = None
from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score, roc_auc_score
from sklearn.utils import shuffle

### Загрузка данных

In [2]:
try:
    df = pd.read_csv('Churn.csv')
except:
    df = pd.read_csv('/datasets/Churn.csv')

### Функция первичного анализа данных

In [3]:
def data_info(data):
    data.info()
    print('')
    print('*'*50)
    display(data.head())
    print('')
    print('*'*50)
    display(data.describe())

### Первичный анализ данных

In [4]:
data_info(df)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB

**************************************************


RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0


**************************************************


RowNumber    CustomerId   CreditScore           Age       Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  9091.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800     4.997690   
std     2886.89568  7.193619e+04     96.653299     10.487806     2.894723   
min        1.00000  1.556570e+07    350.000000     18.000000     0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000     2.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000     5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000     7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000    10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

Первичный анализ данных показал, что имеются пропуски в стольце `Tenure`.

### Проверка на дубликаты

In [5]:
df.duplicated().sum()

0

Проверка на наличие дубликатов показала, что они в данных отсутствуют

### Проверка пропущенных значений

In [6]:
df[df['Tenure'].isna() == True]['Exited'].value_counts()

0    726
1    183
Name: Exited, dtype: int64

In [7]:
df['Tenure'].value_counts()

1.0     952
2.0     950
8.0     933
3.0     928
5.0     927
7.0     925
4.0     885
9.0     882
6.0     881
10.0    446
0.0     382
Name: Tenure, dtype: int64

In [8]:
df['Tenure'] = df['Tenure'].fillna(0)

Пропуски в столбце с данными о длительности пользования банком заполнила нулями для дальнейшего анализа

### Проверка значений столбца `Geography`

In [9]:
df['Geography'].unique()

array(['France', 'Spain', 'Germany'], dtype=object)

Проверила количество уникальных значений столбца `Geography` для дальнейшей обработки

### Вывод

В ходе первичного анализа выявлено:

* Дубли в данных отсутствуют
* Пропуски в данных были в одном столбце `Tenure` - значения, показывющие сколько лет пользователь был клиентом банка
* В столбце страна проживания всего **3** уникальных значения, следовательно можно будет преобразовать этот категориальный признак

Для дальнейшего корректного обучения моделей пропуски заполнила нулями.

## Исследование задачи

### Изучение баланса классов целевого признака

In [10]:
df['Exited'].value_counts()

0    7963
1    2037
Name: Exited, dtype: int64

In [11]:
print('Процентное соотношение классов "клиент остался" {:.0%}, "клиент ушел" {:.0%}'.format(len(df[df['Exited'] == 0])/len(df), 
                                                                          len(df[df['Exited'] == 1])/len(df)))

Процентное соотношение классов "клиент остался" 80%, "клиент ушел" 20%


### Формирование признаков и целевого признака

In [12]:
features = df.drop(['Exited', 'RowNumber', 'CustomerId', 'Surname'], axis=1)
target = df['Exited']
features.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42     2.0       0.00              1   
1          608     Spain  Female   41     1.0   83807.86              1   
2          502    France  Female   42     8.0  159660.80              3   
3          699    France  Female   39     1.0       0.00              2   
4          850     Spain  Female   43     2.0  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  
0          1               1        101348.88  
1          0               1        112542.58  
2          1               0        113931.57  
3          0               0         93826.63  
4          1               1         79084.10

Из признаков для дальнейшего обучения моделей исключила столбцы `RowNumber`, `CustomerId` и `Surname`, т.к. эти данные не влияют на целевой признак.

### Преобразование категориальных признаков, а так же масштабирование признаков

In [13]:
features = pd.get_dummies(features, drop_first=True)

numeric = ['CreditScore', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'EstimatedSalary']
scaler = StandardScaler()
scaler.fit(features[numeric])
features[numeric] = scaler.transform(features[numeric])
features.head()

CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
0    -0.326221  0.293517 -0.817441 -1.225848      -0.911583          1   
1    -0.440036  0.198164 -1.138838  0.117350      -0.911583          0   
2    -1.536794  0.293517  1.110941  1.333053       2.527057          1   
3     0.501521  0.007457 -1.138838 -1.225848       0.807737          0   
4     2.063884  0.388871 -0.817441  0.785728      -0.911583          1   

   IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
0               1         0.021886                  0                0   
1               1         0.216534                  0                1   
2               0         0.240687                  0                0   
3               0        -0.108918                  0                0   
4               1        -0.365276                  0                1   

   Gender_Male  
0            0  
1            0  
2            0  
3            0  
4            0

Для корректного обучения моделей преобразовала категориальные признаки в числовые путем прямого кодирования **One-Hot Encoding** с помощью функции `get_dummies` с аргументом *drop_first*.
Так же произвела **стандартизирование** признаков из-за разного масштаба значений с помощью `StandardScaler`

### Разбивка данных на обучающую, валидационную и тестовую выборки

In [14]:
features_train, features_test, target_train, target_test = train_test_split(
    features, target, test_size=0.2, random_state=12345, stratify=target)
features_train, features_valid, target_train, target_valid = train_test_split(
    features_train, target_train, test_size=0.25, random_state=12345, stratify=target_train)
print('Размер обучающей выборки:', features_train.shape, 
      ', т.е.','{:.1%}'.format(len(features_train)/len(df)), 
      'от общего размера', end='\n')
print('Размер валидационной выборки:', features_valid.shape, 
      ', т.е.','{:.1%}'.format(len(features_valid)/len(df)), 
      'от общего размера', end='\n')
print('Размер тестовой выборки:', features_test.shape, 
      ', т.е.','{:.1%}'.format(len(features_test)/len(df)), 
      'от общего размера', end='\n')

Размер обучающей выборки: (6000, 11) , т.е. 60.0% от общего размера
Размер валидационной выборки: (2000, 11) , т.е. 20.0% от общего размера
Размер тестовой выборки: (2000, 11) , т.е. 20.0% от общего размера


Разделила исходные данные на **обучающую**, **валидационную** и **тестовую** выборки с сохранением баланса классов целевого признака. 

### Функция подбора гиперпараметров для модели дерево решений

In [15]:
def tree(fear_train, target_train, fear_valid, target_valid, random_state, max_depth, class_weight=None):
    best_depth = 0
    best_result_ac = 0
    best_result_re = 0
    best_result_pre = 0
    best_result_f1 = 0
    for depth in range(1, max_depth+1):
        model = DecisionTreeClassifier(max_depth = depth, random_state = random_state, class_weight=class_weight)
        model.fit(fear_train, target_train)
        prediction_valid = model.predict(fear_valid)
        result_ac = accuracy_score(target_valid, prediction_valid)
        result_re = recall_score(target_valid, prediction_valid)
        result_pre = precision_score(target_valid, prediction_valid, zero_division=0)
        result_f1 = f1_score(target_valid, prediction_valid)
        print('max_depth =', depth)
        print('')
        print('Accuracy :{:.3f}'.format(result_ac))
        print('Recall :{:.3f}'.format(result_re))
        print('Precision :{:.3f}'.format(result_pre))
        print('F1 :{:.3f}'.format(result_f1))
        print('-'*20)
        print('')
        if result_f1 > best_result_f1:
            best_depth = depth
            best_result_ac = result_ac
            best_result_pre = result_pre
            best_result_re = result_re
            best_result_f1 = result_f1
            model_tree = model
    print('*'*40)
    print('''Наилучший результат 
    
    Accuracy {:.3f}
    Recall :{:.3f}
    Precision :{:.3f}
    F1 :{:.3f} 
    
    при максимальной глубине дерева {}'''.format(best_result_ac, best_result_re, best_result_pre, best_result_f1, best_depth))
    print('*'*40)
    return model_tree 

### Сохранение модели дерево решений с наилучшими параметрами

In [16]:
model_tree = tree(features_train, target_train, features_valid, target_valid, 12345, 10)

max_depth = 1

Accuracy :0.796
Recall :0.000
Precision :0.000
F1 :0.000
--------------------

max_depth = 2

Accuracy :0.836
Recall :0.428
Precision :0.647
F1 :0.515
--------------------

max_depth = 3

Accuracy :0.835
Recall :0.243
Precision :0.818
F1 :0.375
--------------------

max_depth = 4

Accuracy :0.850
Recall :0.396
Precision :0.745
F1 :0.517
--------------------

max_depth = 5

Accuracy :0.847
Recall :0.322
Precision :0.809
F1 :0.460
--------------------

max_depth = 6

Accuracy :0.854
Recall :0.405
Precision :0.764
F1 :0.530
--------------------

max_depth = 7

Accuracy :0.854
Recall :0.430
Precision :0.745
F1 :0.545
--------------------

max_depth = 8

Accuracy :0.856
Recall :0.494
Precision :0.708
F1 :0.582
--------------------

max_depth = 9

Accuracy :0.847
Recall :0.501
Precision :0.662
F1 :0.571
--------------------

max_depth = 10

Accuracy :0.838
Recall :0.455
Precision :0.642
F1 :0.532
--------------------

****************************************
Наилучший результа

### Функция подбора гиперпараметров для модели слуйчайный лес

In [17]:
def forest(fear_train, target_train, fear_valid, target_valid, random_state, max_depth, n_est, class_weight=None):
    best_depth = 0
    best_est = 0
    best_result_ac = 0
    best_result_re = 0
    best_result_pre = 0
    best_result_f1 = 0
    for est in range(1, n_est+1):
#         print('n_est =', est)
#         print('_'*25)
#         print('')
        for depth in range(1, max_depth+1):
            model = RandomForestClassifier(random_state = random_state, 
                                           max_depth = depth, 
                                           n_estimators = est, 
                                           class_weight=class_weight)
            model.fit(fear_train, target_train)
            prediction_valid = model.predict(fear_valid)
            result_ac = accuracy_score(target_valid, prediction_valid)
            result_re = recall_score(target_valid, prediction_valid)
            result_pre = precision_score(target_valid, prediction_valid, zero_division=0)
            result_f1 = f1_score(target_valid, prediction_valid)
#             print('max_depth =', depth)
#             print('')
#             print('Accuracy :{:.3f}'.format(result_ac))
#             print('Recall :{:.3f}'.format(result_re))
#             print('Precision :{:.3f}'.format(result_pre))
#             print('F1 :{:.3f}'.format(result_f1))
#             print('-'*20)
#             print('')
            if result_f1 > best_result_f1:
                best_depth = depth
                best_est = est
                best_result_ac = result_ac
                best_result_pre = result_pre
                best_result_re = result_re
                best_result_f1 = result_f1
                model_forest = model
#         print('_'*25)
#         print('')
#     print('*'*20)
    print('''Наилучший результат 
    
    Accuracy {:.3f}
    Recall :{:.3f}
    Precision :{:.3f}
    F1 :{:.3f} 
    
при максимальной глубине дерева {} 
и максимальном количестве деревьев {}'''.format(best_result_ac, 
                                                best_result_re, 
                                                best_result_pre, 
                                                best_result_f1, 
                                                best_depth, 
                                                best_est))
    print('*'*40)
    return model_forest 

### Сохранение модели случайный лес с наилучшими гиперпараметрами

In [18]:
model_forest = forest(features_train, target_train, features_valid, target_valid, 12345, 15, 20)

Наилучший результат 
    
    Accuracy 0.864
    Recall :0.452
    Precision :0.790
    F1 :0.575 
    
при максимальной глубине дерева 12 
и максимальном количестве деревьев 17
****************************************


### Функция для подбора гиперпараметров для модели логистическая регрессия

In [19]:
def regression(fear_train, target_train, fear_valid, target_valid, random_state, max_iter, class_weight=None):
    best_result_ac = 0
    best_result_re = 0
    best_result_pre = 0
    best_result_f1 = 0
    best_iter = 0
    for iters in range(100, max_iter+1, 100):
        model = LogisticRegression(random_state = random_state, 
                                   solver='lbfgs', 
                                   max_iter=iters, 
                                   multi_class='multinomial', 
                                   class_weight=class_weight)
        model.fit(fear_train, target_train)
        prediction_valid = model.predict(fear_valid)
        result_ac = accuracy_score(target_valid, prediction_valid)
        result_re = recall_score(target_valid, prediction_valid)
        result_pre = precision_score(target_valid, prediction_valid)
        result_f1 = f1_score(target_valid, prediction_valid)
        print('max_iter =', iters)
        print('')
        print('Accuracy :{:.3f}'.format(result_ac))
        print('Recall :{:.3f}'.format(result_re))
        print('Precision :{:.3f}'.format(result_pre))
        print('F1 :{:.3f}'.format(result_f1))
        print('-'*20)
        print('')

    model_reg = model
    return model_reg

### Сохранение модели логистическая регрессия

In [20]:
model_regression = regression(features_train, target_train, features_valid, target_valid, 12345, 100)

max_iter = 100

Accuracy :0.819
Recall :0.214
Precision :0.674
F1 :0.325
--------------------



### Вывод

В ходе исследования задачи, получила следующие выводы:

1. Баланс классов на исходных данных
    * 80% - клиент остался, целевой признак равен 0
    * 20% - клиент ушел, целевой признак равен 1
2. Для формирования фрейма с признаками исключила 3 столбца с данными, не влияющими на целевой признак
    * `RowNumber` — индекс строки в данных
    * `CustomerId` — уникальный идентификатор клиента
    * `Surname` — фамилия
3. Для корректного обучения моделей необходимо было преобразовать категориальные признаки в численные в столбцах `Gender` и `Geography`
4. Так же из-за разного масштаба в численных значениях мной была применена стандартизация признаков, для корректной работы моделей прогнозирования
5. При игнорирования дисбаланса классов целевого признака имеем следующие показатели:
    * Модель **дерево решений**
        - Accuracy 0.856
        - Recall :0.494
        - Precision :0.708
        - F1 :0.582
    * Модель **случайный лес**
        - Accuracy 0.864
        - Recall :0.452
        - Precision :0.790
        - F1 :0.575
    * Модель **логистическая регрессия**
        - Accuracy :0.819
        - Recall :0.214
        - Precision :0.674
        - F1 :0.325

## Борьба с дисбалансом

### Взвешивание классов

#### Модель дерево решений

In [21]:
model_tree_weight = tree(features_train, target_train, features_valid, target_valid, 12345, 10, 'balanced')

max_depth = 1

Accuracy :0.700
Recall :0.735
Precision :0.378
F1 :0.499
--------------------

max_depth = 2

Accuracy :0.701
Recall :0.757
Precision :0.382
F1 :0.507
--------------------

max_depth = 3

Accuracy :0.806
Recall :0.545
Precision :0.522
F1 :0.534
--------------------

max_depth = 4

Accuracy :0.727
Recall :0.796
Precision :0.411
F1 :0.542
--------------------

max_depth = 5

Accuracy :0.744
Recall :0.799
Precision :0.430
F1 :0.559
--------------------

max_depth = 6

Accuracy :0.760
Recall :0.791
Precision :0.450
F1 :0.573
--------------------

max_depth = 7

Accuracy :0.758
Recall :0.774
Precision :0.446
F1 :0.566
--------------------

max_depth = 8

Accuracy :0.761
Recall :0.744
Precision :0.448
F1 :0.559
--------------------

max_depth = 9

Accuracy :0.750
Recall :0.695
Precision :0.430
F1 :0.531
--------------------

max_depth = 10

Accuracy :0.753
Recall :0.676
Precision :0.432
F1 :0.527
--------------------

****************************************
Наилучший результа

#### Модель случайный лес

In [22]:
model_forest_weight = forest(features_train, target_train, features_valid, target_valid, 12345, 15, 20, 'balanced')

Наилучший результат 
    
    Accuracy 0.834
    Recall :0.703
    Precision :0.575
    F1 :0.633 
    
при максимальной глубине дерева 8 
и максимальном количестве деревьев 16
****************************************


#### Модель логистической регрессии

In [23]:
model_regression_weight = regression(features_train, target_train, features_valid, target_valid, 12345, 100, 'balanced')

max_iter = 100

Accuracy :0.718
Recall :0.722
Precision :0.394
F1 :0.510
--------------------



При использование метода взвешивания классов имеем следующие результаты:

* Модель **дерево решений**
    - Accuracy 0.760
    - Recall :0.791
    - Precision :0.450
    - F1 :0.573
* Модель **случайный лес**
    - Accuracy 0.834
    - Recall :0.703
    - Precision :0.575
    - F1 :0.633
* Модель **логистическая регрессия**
    - Accuracy :0.718
    - Recall :0.722
    - Precision :0.394
    - F1 :0.510

### Увеличение выборки

#### Функция увеличения обучающей выборки

In [24]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

#### Сохранение увеличенных выборок

In [55]:
(features_upsampled_train, 
 target_upsampled_train) = upsample(features_train, 
                                    target_train, 
                                    4)

In [56]:
target_upsampled_train.value_counts()

1    4892
0    4777
Name: Exited, dtype: int64

#### Модель дерево решений

In [34]:
model_tree_up = tree(features_upsampled_train, target_upsampled_train, features_valid, target_valid, 12345, 10)

max_depth = 1

Accuracy :0.700
Recall :0.735
Precision :0.378
F1 :0.499
--------------------

max_depth = 2

Accuracy :0.701
Recall :0.757
Precision :0.382
F1 :0.507
--------------------

max_depth = 3

Accuracy :0.701
Recall :0.757
Precision :0.382
F1 :0.507
--------------------

max_depth = 4

Accuracy :0.727
Recall :0.796
Precision :0.411
F1 :0.542
--------------------

max_depth = 5

Accuracy :0.745
Recall :0.801
Precision :0.431
F1 :0.561
--------------------

max_depth = 6

Accuracy :0.763
Recall :0.789
Precision :0.453
F1 :0.575
--------------------

max_depth = 7

Accuracy :0.759
Recall :0.774
Precision :0.447
F1 :0.567
--------------------

max_depth = 8

Accuracy :0.755
Recall :0.715
Precision :0.438
F1 :0.543
--------------------

max_depth = 9

Accuracy :0.755
Recall :0.703
Precision :0.437
F1 :0.539
--------------------

max_depth = 10

Accuracy :0.756
Recall :0.651
Precision :0.434
F1 :0.521
--------------------

****************************************
Наилучший результа

#### Модель случайный лес

In [35]:
model_forest_up = forest(features_upsampled_train, target_upsampled_train, features_valid, target_valid, 12345, 15, 20)

Наилучший результат 
    
    Accuracy 0.823
    Recall :0.737
    Precision :0.548
    F1 :0.629 
    
при максимальной глубине дерева 9 
и максимальном количестве деревьев 20
****************************************


#### Модель логистическая регрессия

In [36]:
model_regression_up = regression(features_upsampled_train, target_upsampled_train, features_valid, target_valid, 12345, 100)

max_iter = 100

Accuracy :0.712
Recall :0.727
Precision :0.388
F1 :0.506
--------------------



При использовании метода увеличения выборки имеем следующие результаты:

* Модель **дерево решений**
    - Accuracy 0.763
    - Recall :0.789
    - Precision :0.453
    - F1 :0.575
* Модель **случайный лес**
    - Accuracy 0.823
    - Recall :0.737
    - Precision :0.548
    - F1 :0.629
* Модель **логистическая регрессия**
    - Accuracy :0.712
    - Recall :0.727
    - Precision :0.388
    - F1 :0.506

### Уменьшение выборки

#### Функция уменьшения обучающей выборки

In [38]:
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345)
    
    return features_downsampled, target_downsampled

#### Сохранение уменьшенных выборок

In [54]:
(features_downsampled_train, 
 target_downsampled_train) = downsample(features_train, 
                                        target_train, 
                                        0.25)

In [50]:
target_downsampled_train.value_counts()

1    1223
0    1194
Name: Exited, dtype: int64

#### Модель дерево решений

In [51]:
model_tree_down = tree(features_downsampled_train, target_downsampled_train, features_valid, target_valid, 12345, 10)

max_depth = 1

Accuracy :0.668
Recall :0.754
Precision :0.352
F1 :0.480
--------------------

max_depth = 2

Accuracy :0.668
Recall :0.771
Precision :0.355
F1 :0.486
--------------------

max_depth = 3

Accuracy :0.788
Recall :0.555
Precision :0.482
F1 :0.516
--------------------

max_depth = 4

Accuracy :0.703
Recall :0.799
Precision :0.388
F1 :0.523
--------------------

max_depth = 5

Accuracy :0.730
Recall :0.826
Precision :0.417
F1 :0.554
--------------------

max_depth = 6

Accuracy :0.799
Recall :0.722
Precision :0.504
F1 :0.594
--------------------

max_depth = 7

Accuracy :0.737
Recall :0.776
Precision :0.421
F1 :0.546
--------------------

max_depth = 8

Accuracy :0.748
Recall :0.732
Precision :0.430
F1 :0.542
--------------------

max_depth = 9

Accuracy :0.716
Recall :0.730
Precision :0.393
F1 :0.511
--------------------

max_depth = 10

Accuracy :0.725
Recall :0.730
Precision :0.402
F1 :0.519
--------------------

****************************************
Наилучший результа

#### Модель случайный лес

In [52]:
model_forest_down = forest(features_downsampled_train, target_downsampled_train, features_valid, target_valid, 12345, 15, 20)

Наилучший результат 
    
    Accuracy 0.802
    Recall :0.776
    Precision :0.510
    F1 :0.615 
    
при максимальной глубине дерева 8 
и максимальном количестве деревьев 19
****************************************


#### Модель логистическая регрессия

In [53]:
model_regression_down = regression(features_downsampled_train, 
                                   target_downsampled_train, 
                                   features_valid, 
                                   target_valid, 
                                   12345, 
                                   100)

max_iter = 100

Accuracy :0.711
Recall :0.722
Precision :0.387
F1 :0.504
--------------------



При использовании метода уменьшения выборки имеем следующие результаты:

* Модель **дерево решений**
    - Accuracy 0.799
    - Recall :0.722
    - Precision :0.504
    - F1 :0.594 
* Модель **случайный лес**
    - Accuracy 0.802
    - Recall :0.776
    - Precision :0.510
    - F1 :0.615
* Модель **логистическая регрессия**
    - Accuracy :0.711
    - Recall :0.722
    - Precision :0.387
    - F1 :0.504

### Вывод

Наиболее высокие результаты показал метод взвешивания классов:

* Модель **дерево решений**
    - Accuracy 0.760
    - Recall :0.791
    - Precision :0.450
    - F1 :0.573
* Модель **случайный лес**
    - Accuracy 0.834
    - Recall :0.703
    - Precision :0.575
    - F1 :0.633
* Модель **логистическая регрессия**
    - Accuracy :0.718
    - Recall :0.722
    - Precision :0.394
    - F1 :0.510

## Тестирование модели

### Функция подсчета площади под ROC-кривой

In [58]:
def auc_roc(model):
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    print('Площадь под ROC-кривой: {:.3f}'.format(auc_roc))

### Модель дерево решений

In [35]:
predict_tree = model_tree_weight.predict(features_test)
print('Accuracy {:.3f}'.format(accuracy_score(target_test, predict_tree)))
print('Precision {:.3f}'.format(precision_score(target_test, predict_tree)))
print('Recall {:.3f}'.format(recall_score(target_test, predict_tree)))
print('F1 {:.3f}'.format(f1_score(target_test, predict_tree)))
auc_roc(model_tree_weight)

Accuracy 0.760
Precision 0.449
Recall 0.784
F1 0.571
Площадь под ROC-кривой: 0.833


In [59]:
predict_tree = model_tree_up.predict(features_test)
print('Accuracy {:.3f}'.format(accuracy_score(target_test, predict_tree)))
print('Precision {:.3f}'.format(precision_score(target_test, predict_tree)))
print('Recall {:.3f}'.format(recall_score(target_test, predict_tree)))
print('F1 {:.3f}'.format(f1_score(target_test, predict_tree)))
auc_roc(model_tree_up)

Accuracy 0.760
Precision 0.449
Recall 0.781
F1 0.570
Площадь под ROC-кривой: 0.834


### Модель случайный лес

In [36]:
predict_forest = model_forest_weight.predict(features_test)
print('Accuracy {:.3f}'.format(accuracy_score(target_test, predict_forest)))
print('Precision {:.3f}'.format(precision_score(target_test, predict_forest)))
print('Recall {:.3f}'.format(recall_score(target_test, predict_forest)))
print('F1 {:.3f}'.format(f1_score(target_test, predict_forest)))
auc_roc(model_forest_weight)

Accuracy 0.828
Precision 0.561
Recall 0.715
F1 0.629
Площадь под ROC-кривой: 0.863


In [60]:
predict_forest = model_forest_up.predict(features_test)
print('Accuracy {:.3f}'.format(accuracy_score(target_test, predict_forest)))
print('Precision {:.3f}'.format(precision_score(target_test, predict_forest)))
print('Recall {:.3f}'.format(recall_score(target_test, predict_forest)))
print('F1 {:.3f}'.format(f1_score(target_test, predict_forest)))
auc_roc(model_forest_up)

Accuracy 0.822
Precision 0.546
Recall 0.735
F1 0.626
Площадь под ROC-кривой: 0.868


### Модель логистическая регрессия

In [37]:
predict_regression = model_regression_weight.predict(features_test)
print('Accuracy {:.3f}'.format(accuracy_score(target_test, predict_regression)))
print('Precision {:.3f}'.format(precision_score(target_test, predict_regression)))
print('Recall {:.3f}'.format(recall_score(target_test, predict_regression)))
print('F1 {:.3f}'.format(f1_score(target_test, predict_regression)))
auc_roc(model_regression_weight)

Accuracy 0.710
Precision 0.387
Recall 0.730
F1 0.506
Площадь под ROC-кривой: 0.792


### Вывод

При проверке моделей на тестовой выборке с использованием метода взвешивания классов получены следующие результаты:

* Модель **дерево решений**
    - Accuracy 0.760
    - Precision 0.449
    - Recall 0.784
    - F1 0.571
    - Площадь под ROC-кривой: 0.833
* Модель **случайный лес**
    - Accuracy 0.828
    - Precision 0.561
    - Recall 0.715
    - F1 0.629
    - Площадь под ROC-кривой: 0.863
* Модель **логистическая регрессия**
    - Accuracy 0.710
    - Precision 0.387
    - Recall 0.730
    - F1 0.506
    - Площадь под ROC-кривой: 0.792

## Выводы

В ходе исследования были рассмотены три модели классификации:

* Модель **дерево решений**
* Модель **случайный лес**
* Модель **логистическая регрессия**

Так же в ходе исследования были рассмотрены три метода борьбы с дисбалансом классов целевого признака:

* Метод **взвешивание классов**
* Метод **увеличение выборки**
* Метод **уменьшение выборки**

Наилучшим образом на валидационной и тестовой выборках показала себя модель **случайный лес**, получившая следующие результаты:

* Accuracy 0.828
* Precision 0.561
* Recall 0.715
* F1 0.629
* Площадь под ROC-кривой: 0.863

Достигнуто необходимое значение `F1`-меры - 0.63, так же рассчитано значение площади под кривой `auc-roc` для оценки качества модели.